### PR TITLE
CXP-387: check webhook url accessibility

### PR DESCRIPTION
**Pull request by tomglvng**
_As Filips, I want to know if the WebHook URL is accessible._

**OK**

Capture du 2020-09-17 15-46-20

**KO : invalid URL**

Capture du 2020-09-17 15-45-45

**KO: server error**

Capture du 2020-09-17 15-47-01

### CODE REVIEW

File: src/Akeneo/Connectivity/Connection/back/Application/Webhook/Command/CheckWebhookAccessibilityHandler.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        return $checkResponse;
+    }
+
+    /**
+     * @param string $url
+     * @return bool
+     */
+    private function checkUrlFormat(string $url): bool
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Application/Webhook/Command/CheckWebhookAccessibilityHandler.php:69]: https://symfony.com/doc/current/validation/raw_values.html

File: src/Akeneo/Connectivity/Connection/back/Application/Webhook/Command/CheckWebhookAccessibilityHandler.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    /**
+     * @param CheckWebhookAccessibilityCommand $command
+     * @return array
+     */
+    public function handle(CheckWebhookAccessibilityCommand $command): array
+    {
+        if (!$this->checkUrlFormat($command->webhookUrl())) {
+            return [
+                'success' => 'false',
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Application/Webhook/Command/CheckWebhookAccessibilityHandler.php:37]: Boolean

Comment (tomglvng) [reply src/Akeneo/Connectivity/Connection/back/Application/Webhook/Command/CheckWebhookAccessibilityHandler.php:37]: + /domain/Webhook/DTO/UrlReachabilityStatus.php

File: src/Akeneo/Connectivity/Connection/back/Application/Webhook/Command/CheckWebhookAccessibilityHandler.php [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                'message' => 'akeneo_connectivity.connection.webhook.constraint.url.invalid_format',
+            ];
+        }
+
+        $checkResponse = [
+            'success' => 'false',
+            'message' => 'Failed to connect to server',
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Application/Webhook/Command/CheckWebhookAccessibilityHandler.php:44]: Add translation

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Symfony/Resources/translations/jsmessages.en_US.yml [hunk 4]
```diff
@@ -194,17 +194,20 @@ akeneo_connectivity.connection:
...
-
     webhook:
         title: Event subscription (BETA)
         event_subscription: Event subscription (bêta)
         helper:
             message: Your connected app can subscribe to be notified of events in the Akeneo PIM (for example when a product is updated) at a URL you choose.
             link: Learn more about event subscriptions configuration...
+        constraint:
+            url:
+                invalid_format: Invalid url format.
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Symfony/Resources/translations/jsmessages.en_US.yml:205]: See Willy 'wrong_url'

File: src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Application/Webhook/Command/CheckWebhookAccessibilityCommandSpec.php [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    public function it_is_initializable(): void
+    {
+        $this->shouldBeAnInstanceOf(CheckWebhookAccessibilityCommand::class);
+    }
+
+    public function it_returns_the_webhook_url(): void
+    {
+        $this->webhookUrl()->shouldReturn('http://172.17.0.1:8000/webhook');
+    }
+}
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Application/Webhook/Command/CheckWebhookAccessibilityCommandSpec.php:30]: Blank line

File: src/Akeneo/Connectivity/Connection/back/Application/Webhook/Command/CheckWebhookAccessibilityHandler.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        ];
+
+        try {
+
+            $response = $this->client->send(new Request(self::POST, $command->webhookUrl()));
+
+            $checkResponse['success'] = 'true';
+            $checkResponse['message'] = $response->getReasonPhrase();
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Application/Webhook/Command/CheckWebhookAccessibilityHandler.php:52]: TODO : concat code + message in back

File: src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Application/Webhook/Command/CheckWebhookAccessibilityHandlerSpec.php [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    {
+        $checkWebhookAccessibilityCommand = new CheckWebhookAccessibilityCommand('it_is_not_a_good_url_format');
+        $request = new Request($this->getWrappedObject()::POST, $checkWebhookAccessibilityCommand->webhookUrl());
+
+        $client->send($request)->willReturn(new Response(200, [], null, '1.1', 'OK'));
+
+        $this->handle($checkWebhookAccessibilityCommand)->shouldReturn(
+            [
+                'success' => 'false',
+                'message' => 'akeneo_connectivity.connection.webhook.constraint.url.invalid_format',
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/tests/Unit/spec/Application/Webhook/Command/CheckWebhookAccessibilityHandlerSpec.php:59]: update 'wrong_url'

File: src/Akeneo/Connectivity/Connection/back/Application/Webhook/Command/CheckWebhookAccessibilityHandler.php [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+ */
+final class CheckWebhookAccessibilityHandler
+{
+    /** @var string */
+    const POST = 'POST';
+
+    /** @var ClientInterface */
+    private $client;
+
+    public function __construct(ClientInterface $client)
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Application/Webhook/Command/CheckWebhookAccessibilityHandler.php:24]: TODO : add service 'checkAvailability'

File: src/Akeneo/Connectivity/Connection/front/src/webhook/components/EditForm.tsx [hunk 9]
```diff
@@ -15,35 +18,55 @@ type Props = {
...
+    const checkAccessibility = useCheckAccessibility(webhook);
+
+    const [resultTestUrlMessage, setResultTestUrlMessage] = useState<string | null>();
+    const [resultTestUrlCode, setResultTestUrlCode] = useState<number | null>();
+    const [resultTestUrlSuccess, setResultTestUrlSuccess] = useState<string | null>();
+
+    const handleClickTestUrlButton = async () => {
+        const result = await checkAccessibility(getValues('url'));
+
+        if (isOk(result)) {
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/front/src/webhook/components/EditForm.tsx:31]: Set Object

File: src/Akeneo/Connectivity/Connection/front/src/webhook/components/EditForm.tsx [hunk 10]
```diff
@@ -15,35 +18,55 @@ type Props = {
...
-                    ref={register}
-                />
+                <>
+                    <FormInput
+                        type='url'
+                        name='url'
+                        defaultValue={webhook.url ? webhook.url : undefined}
+                        ref={register}
+                    />
+                    <TestUrlButton onClick={handleClickTestUrlButton} />
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/front/src/webhook/components/EditForm.tsx:56]: handleTestUrlButton

File: src/Akeneo/Connectivity/Connection/front/src/webhook/components/EditForm.tsx [hunk 11]
```diff
@@ -60,3 +83,56 @@ export const CredentialList = styled.div`
         border-top: 1px solid ${({theme}) => theme.color.grey80};
     }
 `;
+
+const TestUrlButton: FC<{onClick: () => void}> = ({onClick}) => {
+    return (
+        <GreyLightButtonContainer>
+            <GreyLightButton onClick={onClick} classNames={['AknButton--apply']}>
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/front/src/webhook/components/EditForm.tsx:90]: Ask Sophie for similar button

File: src/Akeneo/Connectivity/Connection/front/src/webhook/components/EditForm.tsx [hunk 12]
```diff
@@ -60,3 +83,56 @@ export const CredentialList = styled.div`
...
+    background-size: 20px;
+    background-position: top left;
+`;
+
+const KoStatus = styled.span`
+    display: flex;
+    align-items: baseline;
+    margin-top: -14px;
+    margin-bottom: 18px;
+    background: url('/bundles/pimui/images/icon-danger.svg') no-repeat left center;
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/front/src/webhook/components/EditForm.tsx:129]: create svg

File: src/Akeneo/Connectivity/Connection/front/src/webhook/model/WebhookAccessibility.ts [hunk 13]
```diff
@@ -0,0 +1,5 @@
+export type WebhookAccessibility = {
+    success: string | null | undefined;
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/front/src/webhook/model/WebhookAccessibility.ts:2]: remove undefined

File: src/Akeneo/Connectivity/Connection/front/src/webhook/hooks/api/use-webhook-check-accessibility.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
+import {fetchResult} from '../../../shared/fetch-result';
+import {useRoute} from '../../../shared/router';
+import {Webhook} from '../../model/Webhook';
+import {WebhookAccessibility} from '../../model/WebhookAccessibility';
+
+const useCheckAccessibility = (webhook: Webhook) => {
+    const code = webhook.connectionCode;
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/front/src/webhook/hooks/api/use-webhook-check-accessibility.ts:7]: connectionCode not webhook

File: src/Akeneo/Connectivity/Connection/back/Application/Webhook/Service/WebhookReachabilityChecker.php [hunk 15]
```diff
@@ -0,0 +1,102 @@
+<?php
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Application/Webhook/Service/WebhookReachabilityChecker.php:1]: Move to Infra + Add interface in Application

File: src/Akeneo/Connectivity/Connection/back/Application/Webhook/Service/WebhookReachabilityChecker.php [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+                );
+            }
+        }
+    }
+
+    /**
+     * @param string $url
+     * @return bool
+     */
+    private function checkUrlFormat(string $url): bool
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Application/Webhook/Service/WebhookReachabilityChecker.php:86]: To test : return message violation

File: src/Akeneo/Connectivity/Connection/back/Application/Webhook/Service/WebhookReachabilityChecker.php [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+            $response = $this->client->send(new Request(self::POST, $url));
+
+            return new UrlReachabilityStatus(
+                true,
+                $this->buildMessage($response->getReasonPhrase(), $response->getStatusCode())
+            );
+        } catch (GuzzleException $e) {
+            if ($e instanceof RequestException && $e->hasResponse()) {
+
+                /** @var  ResponseInterface */
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/Application/Webhook/Service/WebhookReachabilityChecker.php:66]: remove space

File: src/Akeneo/Connectivity/Connection/back/tests/.php_cd.php [hunk 18]
```diff
@@ -26,6 +26,14 @@
             'FOS\RestBundle\Serializer\Serializer',
 
             'Akeneo\Platform\Component\EventQueue\BusinessEventInterface',
+
```
Comment (tomglvng) [inline src/Akeneo/Connectivity/Connection/back/tests/.php_cd.php:29]: To remove

File: src/Akeneo/Connectivity/Connection/back/Application/Webhook/Service/WebhookReachabilityChecker.php [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+
+    /**
+     * @param string $reasonPhrase
+     * @param int|null $code
+     * @return string
+     */
+    private function buildMessage(string $reasonPhrase, int $code = null): string
+    {
+        return $code ? sprintf("%s %s", $code, $reasonPhrase) : $reasonPhrase;
+    }
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/Application/Webhook/Service/WebhookReachabilityChecker.php:101]: re-reading this now, it does seem overkill to split this in his own function

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Symfony/Resources/config/routing/webhook.yml [hunk 20]
```diff
@@ -4,3 +4,10 @@ get:
   methods: [GET]
   requirements:
     code: '[a-zA-Z0-9_]+'
+
+check_reachability:
+  path: '/{code}/webhook/check_reachability'
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Symfony/Resources/config/routing/webhook.yml:9]:
```suggestion
  path: '/{code}/webhook/check-reachability'
```
We use `-` for url

**Review by pchasle — APPROVED** [review]

**Review by pierrejolly — APPROVED** [review]

**Pull request merged**